flink-core/src/test/java/org/apache/flink/core/memory/ByteBufferUtilsTest.java

```diff
@@ -0,0 +1,195 @@
+		testCompareTo(false, true, true);
+	}
+
+	@Test
+	public void testCompareHeapBBToHeapBB() {
+		testCompareTo(false, true, false);
+	}
+
+	private void testCompareTo(boolean isLeftBBDirect, boolean isRightBuffer, boolean isRightDirect) {
+		testEquals(isLeftBBDirect, isRightBuffer, isRightDirect);
+		testLessThan(isLeftBBDirect, isRightBuffer, isRightDirect);
+		testGreaterThan(isLeftBBDirect, isRightBuffer, isRightDirect);
+	}
+
+	private void testEquals(boolean isLeftBBDirect, boolean isRightBuffer, boolean isRightDirect) {
+		byte[] leftBufferBytes = new byte[]{'a', 'b', 'c', 'd', 'e'};
+		byte[] rightBufferBytes = new byte[]{'b', 'c', 'd', 'e', 'f'};
+		ByteBuffer left = isLeftBBDirect
+			? ByteBuffer.allocateDirect(leftBufferBytes.length).put(leftBufferBytes)
+			: ByteBuffer.wrap(leftBufferBytes);
+		ByteBuffer right = null;
+		if (isRightBuffer) {
+			right = isRightDirect
+				? ByteBuffer.allocateDirect(rightBufferBytes.length).put(rightBufferBytes)
+				: ByteBuffer.wrap(rightBufferBytes);
+		}
+		if (right != null) {
+			Assert.assertThat(
+				ByteBufferUtils.compareTo(left, 1, 4, right, 0, 4),
+				is(0));
+		} else {
+			Assert.assertThat(
+				ByteBufferUtils.compareTo(left, 1, 4, rightBufferBytes, 0, 4),
+				is(0));
+		}
+	}
+
+	private void testLessThan(boolean isLeftBBDirect, boolean isRightBuffer, boolean isRightDirect) {
+		byte[] leftBufferBytes = new byte[]{'a', 'b', 'c', 'd', 'e'};
+		byte[] rightBufferBytes = new byte[]{'b', 'c', 'd', 'e', 'f'};
+		ByteBuffer left = isLeftBBDirect
+			? ByteBuffer.allocateDirect(leftBufferBytes.length).put(leftBufferBytes)
+			: ByteBuffer.wrap(leftBufferBytes);
+		ByteBuffer right = null;
+		if (isRightBuffer) {
+			right = isRightDirect
+				? ByteBuffer.allocateDirect(rightBufferBytes.length).put(rightBufferBytes)
+				: ByteBuffer.wrap(rightBufferBytes);
+		}
+		if (right != null) {
+			Assert.assertThat(
+				ByteBufferUtils.compareTo(left, 1, 4, right, 1, 4),
+				lessThan(0));
+		} else {
+			Assert.assertThat(
+				ByteBufferUtils.compareTo(left, 1, 4, rightBufferBytes, 1, 4),
+				lessThan(0));
+		}
+	}
+
+	private void testGreaterThan(boolean isLeftBBDirect, boolean isRightBuffer, boolean isRightDirect) {
+		byte[] leftBufferBytes = new byte[]{'a', 'b', 'c', 'd', 'e'};
+		byte[] rightBufferBytes = new byte[]{'b', 'c', 'd', 'e', 'f'};
+		ByteBuffer left = isLeftBBDirect
+			? ByteBuffer.allocateDirect(leftBufferBytes.length).put(leftBufferBytes)
+			: ByteBuffer.wrap(leftBufferBytes);
+		ByteBuffer right = null;
+		if (isRightBuffer) {
+			right = isRightDirect
+				? ByteBuffer.allocateDirect(rightBufferBytes.length).put(rightBufferBytes)
+				: ByteBuffer.wrap(rightBufferBytes);
+		}
+		if (right != null) {
+			Assert.assertThat(
+				ByteBufferUtils.compareTo(left, 1, 4, right, 0, 3),
+				greaterThan(0));
+		} else {
+			Assert.assertThat(
+				ByteBufferUtils.compareTo(left, 1, 4, rightBufferBytes, 0, 3),
+				greaterThan(0));
+		}
+	}
```

Comment:
We still cannot simplify it to one line statement since `Either` requires additional check, but will simplify the test logic relatively with these hints.